### ISSUE TITLE
[QUESTION] Expose ports from docker container to host

### ISSUE THREAD

**Issue by balta2ar** [nteract/hydrogen#367]
Hello. I'm trying to run custom kernel inside a docker container. I've managed to run jupyter kernel inside the docker according to the instruction, and I can connect to it using `jupyter console --existing connection.json` if I run it inside the container, but I cannot connect to it if I run it on the host system:
```
$ jupyter console --existing connection.json
Traceback (most recent call last):
  File "/home/ybochkarev/miniconda/envs/cs/bin/jupyter-console", line 11, in <module>
    sys.exit(main())
  File "/home/ybochkarev/miniconda/envs/cs/lib/python2.7/site-packages/jupyter_core/application.py", line 267, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/home/ybochkarev/miniconda/envs/cs/lib/python2.7/site-packages/traitlets/config/application.py", line 595, in launch_instance
    app.initialize(argv)
  File "<decorator-gen-113>", line 2, in initialize
  File "/home/ybochkarev/miniconda/envs/cs/lib/python2.7/site-packages/traitlets/config/application.py", line 74, in catch_config_error
    return method(app, *args, **kwargs)
  File "/home/ybochkarev/miniconda/envs/cs/lib/python2.7/site-packages/jupyter_console/app.py", line 137, in initialize
    self.init_shell()
  File "/home/ybochkarev/miniconda/envs/cs/lib/python2.7/site-packages/jupyter_console/app.py", line 110, in init_shell
    client=self.kernel_client,
  File "/home/ybochkarev/miniconda/envs/cs/lib/python2.7/site-packages/traitlets/config/configurable.py", line 412, in instance
    inst = cls(*args, **kwargs)
  File "/home/ybochkarev/miniconda/envs/cs/lib/python2.7/site-packages/jupyter_console/ptshell.py", line 251, in __init__
    self.init_kernel_info()
  File "/home/ybochkarev/miniconda/envs/cs/lib/python2.7/site-packages/jupyter_console/ptshell.py", line 305, in init_kernel_info
    raise RuntimeError("Kernel didn't respond to kernel_info_request")
RuntimeError: Kernel didn't respond to kernel_info_request
```

Ports that are mentioned in connection.json should probably be somehow exposed to host. Could you please advice an appropriate way to do that considering that they are probably randomly assigned? Thank you.

EDIT: I didn't mention that explicitly, but hydrogen does not work in that configuration. It stucks on the message "Python 2: starting" in the status bar. That's why I decided to simplify things and try jupyter console first.

**Comment by n-riesco** [comment]
What version of Hydrogen are you using?

This functionality is broken in master, but it should be working in the latest release (Hydrogen v0.11.0).

**Comment by n-riesco** [comment]
This works for me in Hydrogen v0.11.0

```
$ ls
Dockerfile  test.py

$ cat Dockerfile 
FROM python:2.7

RUN pip install ipykernel
RUN echo "alias hydrokernel='python -m ipykernel "'--ip=$(hostname -I)'" -f /tmp/hydrogen/connection.json'" >> /etc/bash.bashrc

$ docker build -t python-docker .
Sending build context to Docker daemon 3.072 kB
Step 1 : FROM python:2.7
 ---> a047e3d0ae2b
Step 2 : RUN pip install ipykernel
 ---> Using cache
 ---> fc5b50bb931f
Step 3 : RUN echo "alias hydrokernel='python -m ipykernel "'--ip=$(hostname -I)'" -f /tmp/hydrogen/connection.json'" >> /etc/bash.bashrc
 ---> Using cache
 ---> b9ac2a3aa226
Successfully built b9ac2a3aa226

$ docker rm python-docker 
python-docker

$ docker run -it --name python-docker -v $(pwd)/hydrogen:/tmp/hydrogen python-docker bash
root@593ea989f549:/# hydrokernel 
NOTE: When using the `ipython kernel` entry point, Ctrl-C will not work.

To exit, you will have to explicitly quit this process, by either sending
"quit" from a client, or using Ctrl-\ in UNIX-like environments.

To read more about this, see https://github.com/ipython/ipython/issues/2049


To connect another client to this kernel, use:
    --existing /tmp/hydrogen/connection.json
```

Then open the file `test.py` and you should be able to run the following:

```python
import json
json.load(open("/tmp/hydrogen/connection.json"))
```

that produced this output in my case:

```python
{u'control_port': 40414,
 u'hb_port': 35573,
 u'iopub_port': 37976,
 u'ip': u'172.17.0.2',
 u'kernel_name': u'',
 u'key': u'c94a9f93-b5b5-46a0-9e3b-b202aebe1c63',
 u'shell_port': 33786,
 u'signature_scheme': u'hmac-sha256',
 u'stdin_port': 38759,
 u'transport': u'tcp'}
```

**Comment by balta2ar** [comment]
I don't really think it's related to hydrogen, it was rather documentation/docker question.

In fact, my configuration is slightly different. I'm using docker as a provider in vagrant. For some reason, `$(hostname -I)` when run in a container shows correct IP (172.17.0.2 in my case), but when I run command from README.md `python -m ipykernel "'--ip=$(hostname -I)'" -f /tmp/hydrogen/connection.json`, 127.0.0.1 gets written into connection.json. I have no ideas why. And apparently the kernel binds to 127.0.0.1 interface and is not available from the outside.

I looked at #315 and it helped. When I manually specified the IP, it worked and I could connect to the kernel from host. I also changed network_type to private_network:
``` ruby
  config.vm.network "private_network"
```
I'm not sure if it's important.

Feel free to close this issue as it is superseded by #315. Maybe it's worth digging into why that shell expansion/eval didn't work.

@quintona Have you had any similar issues setting up vagrant?

**Comment by n-riesco** [comment]
Do you have a minimum Vagrantfile I can use to reproduce the issue?

**Comment by balta2ar** [comment]
I think my mistake is that I copied command line from Dockerfile directly and I was running literally this:
``` bash
python -m ipykernel "'--ip=$(hostname -I)'" -f ./connection2.json
```
Notice single quotes. After I removed them and ran as follows, I found correct IP in json:
``` bash
python -m ipykernel "--ip=$(hostname -I)" -f ./connection2.json
```
Though with this notation one extra space is appended to IP address for some reason:
``` json
"ip": "172.17.0.2 ",
```
Again, when I removed it manually, connection went fine. Sorry I have disturbed you about that really novice mistake.
